### PR TITLE
feat: signless execution of eth txs

### PR DESCRIPTION
**Pull request by keruch**
## Description

Allows granting access to execute `MsgEthereumTx` on other's behalf.
ADR: https://www.notion.so/dymension/Extend-EVM-for-Authz-compatible-15ba4a51f86a800492a9d5bd946ab2e5

----

Closes #XXX

### CODE REVIEW

File: x/evm/types/msg.go [hunk 1]
```diff
@@ -382,6 +415,32 @@ func (msg *MsgEthereumTx) BuildTx(b client.TxBuilder, evmDenom string) (signing.
 	return tx, nil
 }
 
+// TxAsMessage ia a modified version of ethtypes.AsMessage that allows to set from address explicitly.
+func TxAsMessage(tx *ethtypes.Transaction, baseFee *big.Int, from common.Address) ethtypes.Message {
```
Comment (keruch) [inline x/evm/types/msg.go:419]: it does the same as `MsgEthereumTx` but does not derive the sender address from the signer and allows to specify from explicitly.

File: x/evm/types/msg.go [hunk 2]
```diff
@@ -314,9 +330,26 @@ func (msg MsgEthereumTx) AsTransaction() *ethtypes.Transaction {
 	return ethtypes.NewTx(txData.AsEthereumData())
 }
 
-// AsMessage creates an Ethereum core.Message from the msg fields
+// AsMessage creates an Ethereum core.Message from the msg fields. If the msg is impersonated, ie
+// has non-empty OnBehalf, then From is replaced with it. Otherwise, the from address is derived
+// from the signature.
 func (msg MsgEthereumTx) AsMessage(signer ethtypes.Signer, baseFee *big.Int) (core.Message, error) {
```
Comment (keruch) [inline x/evm/types/msg.go:336]: this method is important as it replaces `from` in core eth msg. it should be used everywhere instead of `ethtypes.Transaction.AsMessage()`.

File: rpc/types/utils.go [hunk 3]
```diff
@@ -184,22 +189,30 @@ func NewRPCTransaction(
...
 		signer = ethtypes.HomesteadSigner{}
 	}
-	from, _ := ethtypes.Sender(signer, tx) // #nosec G703
+	originalFrom, _ := ethtypes.Sender(signer, tx) // #nosec G703
+
+	from := originalFrom
+	if onBehalf != nil {
+		// Replace the sender address if the transaction is impersonated
+		from = *onBehalf
+	}
```
Comment (keruch) [inline rpc/types/utils.go:198]: this part is important as it replaces `from` in all the `eth_getTransactionBy*` RPC calls.

File: x/evm/keeper/grpc_query.go [hunk 4]
```diff
@@ -507,10 +509,15 @@ func (k Keeper) TraceBlock(c context.Context, req *types.QueryTraceBlockRequest)
 	txConfig := statedb.NewEmptyTxConfig(common.BytesToHash(ctx.HeaderHash().Bytes()))
 	for i, tx := range req.Txs {
 		result := types.TxTraceResult{}
-		ethTx := tx.AsTransaction()
-		txConfig.TxHash = ethTx.Hash()
+		txConfig.TxHash = tx.AsTransaction().Hash()
+		coreMsg, err := tx.AsMessage(signer, cfg.BaseFee)
```
Comment (keruch) [inline x/evm/keeper/grpc_query.go:513]: we need to use `MsgEthereumTx.AsMessage` instead of `ethtypes.Transaction.AsMessage` so we can replace the sender

File: x/evm/keeper/grpc_query.go [hunk 5]
```diff
@@ -426,22 +426,24 @@ func (k Keeper) TraceTx(c context.Context, req *types.QueryTraceTxRequest) (*typ
 
 	txConfig := statedb.NewEmptyTxConfig(common.BytesToHash(ctx.HeaderHash().Bytes()))
 	for i, tx := range req.Predecessors {
-		ethTx := tx.AsTransaction()
-		msg, err := ethTx.AsMessage(signer, cfg.BaseFee)
+		coreMsg, err := tx.AsMessage(signer, cfg.BaseFee)
```
Comment (keruch) [inline x/evm/keeper/grpc_query.go:429]: we need to use `MsgEthereumTx.AsMessage` instead of `ethtypes.Transaction.AsMessage` so we can replace the sender

File: rpc/backend/call_tx.go [hunk 6]
```diff
@@ -244,8 +245,9 @@ func (b *Backend) SetTxDefaults(args evmtypes.TransactionArgs) (evmtypes.Transac
 	}
 	if args.Nonce == nil {
 		// get the nonce from the account retriever
-		// ignore error in case tge account doesn't exist yet
-		nonce, _ := b.getAccountNonce(*args.From, true, 0, b.logger) // #nosec G703s
+		// ignore error in case the account doesn't exist yet
+		// the account may be the original sender or the granter
+		nonce, _ := b.getAccountNonce(args.GetFrom(), true, 0, b.logger) // #nosec G703s
```
Comment (keruch) [inline rpc/backend/call_tx.go]: use the granter's nonce in the tx by default if `OnBehalf` is specified

File: x/evm/types/msg.go [hunk 7]
```diff
@@ -295,13 +298,26 @@ func (msg MsgEthereumTx) GetEffectiveFee(baseFee *big.Int) *big.Int {
 }
 
 // GetFrom loads the ethereum sender address from the sigcache and returns an
-// sdk.AccAddress from its bytes
+// sdk.AccAddress from its bytes. If the sender is not found, it returns nil.
+// If the message has an OnBehalf field, it returns the address from that field.
 func (msg *MsgEthereumTx) GetFrom() sdk.AccAddress {
```
Comment (VictorTrustyDev) [inline x/evm/types/msg.go]: I think should rename `GetFrom` method or create another with name reflect the logic like actual name or smt because this really confused when reading.

File: app/ante/evm/eth.go [hunk 8]
```diff
@@ -301,14 +303,10 @@ func (ctd CanTransferDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate
 		}
 
 		baseFee := ctd.evmKeeper.GetBaseFee(ctx, ethCfg)
-
-		coreMsg, err := msgEthTx.AsMessage(signer, baseFee)
```
Comment (VictorTrustyDev) [inline app/ante/evm/eth.go:305]: as you already updated `msgEthTx.AsMessage` so imo should keep using it instead of this long and cause unnecessary change diff
```go
from := common.BytesToAddress(msgEthTx.GetFrom())
coreMsg := evmtypes.TxAsMessage(msgEthTx.AsTransaction(), baseFee, from)
```

File: app/ante/evm/eth.go [hunk 9]
```diff
@@ -337,9 +335,13 @@ func (ctd CanTransferDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate
...
 		evm := ctd.evmKeeper.NewEVM(ctx, coreMsg, cfg, evmtypes.NewNoOpTracer(), stateDB)
 
+		// Sender address should be in the tx cache from the previous AnteHandle call.
+		// From may be either actual from address or the granter address.
+		sender := common.BytesToAddress(msgEthTx.GetFrom())
+
 		// check that caller has enough balance to cover asset transfer for **topmost** call
 		// NOTE: here the gas consumed is from the context with the infinite gas meter
-		if coreMsg.Value().Sign() > 0 && !evm.Context.CanTransfer(stateDB, coreMsg.From(), coreMsg.Value()) {
+		if coreMsg.Value().Sign() > 0 && !evm.Context.CanTransfer(stateDB, sender, coreMsg.Value()) {
```
Comment (VictorTrustyDev) [inline app/ante/evm/eth.go]: you already replaced the `from` in the core message so imo no need unnecessary change diff like this

File: app/ante/evm/eth.go [hunk 10]
```diff
@@ -366,7 +368,8 @@ func NewEthIncrementSenderSequenceDecorator(ak evmtypes.AccountKeeper) EthIncrem
 
 // AnteHandle handles incrementing the sequence of the signer (i.e. sender). If the transaction is a
 // contract creation, the nonce will be incremented during the transaction execution and not within
-// this AnteHandler decorator.
+// this AnteHandler decorator. If the transaction is executed on behalf, increase the sequence for
+// both granter and grantee.
```
Comment (VictorTrustyDev) [inline app/ante/evm/eth.go]:
```suggestion
// this AnteHandler decorator.
// If the transaction is executed on behalf, increase the sequence for granter only.
```

File: app/ante/evm/eth.go [hunk 11]
```diff
@@ -379,31 +382,43 @@ func (issd EthIncrementSenderSequenceDecorator) AnteHandle(ctx sdk.Context, tx s
...
-		if acc == nil {
-			return ctx, errorsmod.Wrapf(
-				errortypes.ErrUnknownAddress,
-				"account %s is nil", common.BytesToAddress(msgEthTx.GetFrom().Bytes()),
-			)
-		}
-		nonce := acc.GetSequence()
+		// Sender address should be in the tx cache from the previous AnteHandle call.
+		// From may be either actual from address or the granter address.
+		sender := common.BytesToAddress(msgEthTx.GetFrom())
```
Comment (VictorTrustyDev) [inline app/ante/evm/eth.go]: so already updated the `msgEthTx.GetFrom` so I guess the entire change for this decorator is not needed?

File: x/evm/types/tx_args.go [hunk 12]
```diff
@@ -35,6 +35,7 @@ import (
 type TransactionArgs struct {
 	From                 *common.Address `json:"from"`
 	To                   *common.Address `json:"to"`
+	OnBehalf             *common.Address `json:"on_behalf"`
```
Comment (VictorTrustyDev) [inline x/evm/types/tx_args.go]:
```suggestion
	OnBehalf             *common.Address `json:"onBehalf"`
```
Naming doesn't match with others

**Review by VictorTrustyDev — CHANGES_REQUESTED** [review]
Dear Kingrill,
This is awesome.

Just some nit.

File: x/evm/types/msg.go (hunk 7 above)
Comment (keruch) [reply x/evm/types/msg.go]: sure, renamed it to `GetEffectiveSender`. hope it makes more sense!

File: rpc/backend/sign_tx.go [hunk 13]
```diff
@@ -35,10 +35,12 @@ import (
...
 func (b *Backend) SendTransaction(args evmtypes.TransactionArgs) (common.Hash, error) {
-	// Look up the wallet containing the requested signer
-	_, err := b.clientCtx.Keyring.KeyByAddress(sdk.AccAddress(args.GetFrom().Bytes()))
+	// Look up the wallet containing the requested signer. We need to retrieve the original
+	// sender of the transaction (not the granter) to sign the transaction. We do not know
+	// the granter's private key thus won't find it in the keyring.
+	_, err := b.clientCtx.Keyring.KeyByAddress(sdk.AccAddress(args.GetOriginalFrom().Bytes()))
 	if err != nil {
-		b.logger.Error("failed to find key in keyring", "address", args.GetFrom(), "error", err.Error())
+		b.logger.Error("failed to find key in keyring", "address", args.GetEffectiveSender(), "error", err.Error())
```
Comment (VictorTrustyDev) [inline rpc/backend/sign_tx.go]:
```suggestion
		b.logger.Error("failed to find key in keyring", "address", args.GetOriginalFrom(), "error", err.Error())
```

**Review by VictorTrustyDev — CHANGES_REQUESTED** [review]
